ext/liquid_c/vm_assembler.c

```diff
@@ -73,6 +74,113 @@ void vm_assembler_gc_mark(vm_assembler_t *code)
     }
 }
 
+VALUE vm_assembler_disassemble(vm_assembler_t *code)
+{
+    const size_t *const_ptr = (size_t *)code->constants.data;
+    const uint8_t *start_ip = code->instructions.data;
+    const uint8_t *ip = start_ip;
+    const uint8_t *end_ip = code->instructions.data_end;
+    VALUE output = rb_str_buf_new(32);
+    while (ip < end_ip) {
+        rb_str_catf(output, "0x%04lx: ", ip - start_ip);
+        switch (*ip) {
+            case OP_LEAVE:
+                rb_str_catf(output, "leave\n");
+                break;
+
+            case OP_POP_WRITE:
+                rb_str_catf(output, "pop_write\n");
+                break;
+
+            case OP_PUSH_NIL:
+                rb_str_catf(output, "push_nil\n");
+                break;
+
+            case OP_PUSH_TRUE:
+                rb_str_catf(output, "push_true\n");
+                break;
+
+            case OP_PUSH_FALSE:
+                rb_str_catf(output, "push_false\n");
+                break;
+
+            case OP_FIND_VAR:
+                rb_str_catf(output, "find_var\n");
+                break;
+
+            case OP_LOOKUP_KEY:
+                rb_str_catf(output, "lookup_key\n");
+                break;
+
+            case OP_NEW_INT_RANGE:
+                rb_str_catf(output, "new_int_range\n");
+                break;
+
+            case OP_HASH_NEW:
+                rb_str_catf(output, "hash_new(%u)\n", ip[1]);
+                break;
+
+            case OP_PUSH_INT8:
+                rb_str_catf(output, "push_int8(%u)\n", ip[1]);
+                break;
+
+            case OP_PUSH_INT16:
+            {
+                int num = *(int8_t *)ip++;
+                num = (num << 8) | ip[1];
+                rb_str_catf(output, "push_int16(%u)\n", num);
+                break;
+            }
+
+            case OP_RENDER_VARIABLE_RESCUE:
+            {
+                unsigned int line_number = decode_node_line_number(ip + 1);
+                rb_str_catf(output, "render_variable_rescue(line_number: %u)\n", line_number);
+                break;
+            }
+
+            case OP_WRITE_RAW:
+            {
+                const char *text = (const char *)const_ptr[0];
+                size_t size = const_ptr[1];
+                VALUE string = rb_enc_str_new(text, size, utf8_encoding);
+                rb_str_catf(output, "write_raw(%+"PRIsVALUE")\n", string);
```

Comment:
Can we do this?
```suggestion
                rb_str_catf(output, "write_raw(\"%.*s\")\n", (int)size, text);
```
It would avoid allocating a string.

Reply:
I'm not really concerned about the performance of disassemble, since it is provided for testing and debugging.  Your suggested change wouldn't escape special characters, such as newlines or quote characters.

Reply:
Ah yeah I didn't think about newlines/Special characters. Makes sense, thanks for clarifying!